### DISCUSSION TITLE
Mandatory use of designated prefixes

### DISCUSSION THREAD

**Discussion by MatthiasHainz-ceg** [ConnectingEurope/eInvoicing-EN16931#402] [Q&A] [answered]
Hello all,

I came across the question, if the defined namespace prefixes (ram, qdt, udt,...) must be used in each Instance of an Invoice when using the CII Syntax. My validations against the schematron are all succesfull, even after replacing the ram-prefix with an lll-prefix.

Taking a look at [XML Naming and Design Rules](https://unece.org/DAM/cefact/xml/UNCEFACT+XML+NDR+V3p0.pdf) I understood that there should be fixed 3 letter prefixes for each data package which are to be used during schema design.

The question arose, as we came across an instance where <ram:ID> was written as <ID xmlns="urn:un:uncefact:data:standard:ReusableAggregateBusinessInformationEntity:100>. As this notation is missing the prefix, we were not sure about its conformity.

To summarize, the question is:
Is the use of the defined prefixes mandatory wherever it is applicable? Ans if so, why does the validation not catch these scenarios?

**Comment by phax** [comment, marked as answer]
Hi @MatthiasHainz-ceg , the usage of specific XML prefixes is not mandatory in the context of EN16931 invoices.
The XML prefix is only a shortcut to represent long namespace URIs in serialized versions of XML.
Please see https://www.w3.org/TR/xml-names/ for the official specification.

**Comment by AndreasPvd** [comment]
Hi @MatthiasHainz-ceg  and @phax,

What Philip writes here is not entirely true. Although XML itself allows different representation of namespaces, the CrossIndustryInvoice does not(!). CEN/TS 16931-2 states that CrossIndustryInvoice is one of the allowed syntaxes. One important requirement is that CrossIndustryInvoice is an international standard. And the standard follows a set of rules.

It is true, that EN16931 is just semantics itself and neither the syntax list (CEN/TS 16931-2) nor the syntax bindings (CEN/TS 16931-3-*) add any additional constraints. 

As a result the rules of the international standards UBL and CII must be followed.

For CrossIndustryInvoice D16B the [XMLNamingAndDesignRulesV2.1.0.pdf ](https://unece.org/DAM/cefact/xml/XML-Naming-And-Design-Rules-V2_1.pdf) is defining the standard. In Cahpter 7.1 it clearly states that the prefixes rsm, ram, qdt and udt MUST be used. It is mandatory.

The current version D23A is based on https://unece.org/sites/default/files/2023-10/XMLNamingAndDesignRulesV2.1.1.pdf which has the same set of rules in the same chapter.

It forbids to use the long namespace representation.

As a result the answer must be Yes, usage of namespace prefixes using CII is mandatory in the context of EN16931.

Br,
Andreas

**Reply by AndreasPvd** [reply]
The XML Naming and Design Rules is in this context restricting the possibilities defined by the XML standard.
This is very common. If only the XML standard would rule, you could create any syntax on your own.

UBL and CII are subsets of XML and add special rule sets.

It is like a CIUS of EN16931. Although EN16931 has got many optional elements, a CIUS can and is intended to restrict the use of them or the supported values...

**Reply by phax** [reply]
@AndreasPvd  thanks for the clarifications regarding CII.
However, chapter 7 is labeled "XML Schema Modules" - to my understanding, the naming constraints mentioned in chapter 7 only apply to the creation of XML Schemas and not for instances following such a Schema

**Reply by MatthiasHainz-ceg** [reply]
>  to my understanding, the naming constraints mentioned in chapter 7 only apply to the creation of XML Schemas and not for instances following such a Schema

That was also my understanding. Hence the question how it is handled in the Instances and if the rules are that strict, why isn't there any validation against it?

**Reply by AndreasPvd** [reply]
About validation -> To my understanding this is simply not possible using XSD but can only be achieved using schematron. At least a reason but I a agree not necessary a good one...

But Chapter 8 references chapter 7 to be used for instances as well...
"It should represent all truncated tag names as described in section 7."

**Reply by MartinForsberg-Ecru** [reply]
I agree (edit: I agree with philip), the XML Naming and Design Rules are for the creation of XSDs. From the UN/CEFACT NDR Introduction chapter "This UN/CEFACT – XML Naming and Design Rules Technical Specification describes and specifies the rules and guidelines that will be applied by UN/CEFACT when developing XML schema"

**Reply by MartinForsberg-Ecru** [reply]
I read "truncated tag name" as the truncation of the BIE-name into an XML Element name.

**Reply by AndreasPvd** [reply]
Good point, thanks Martin.

**Reply by MatthiasHainz-ceg** [reply]
> About validation -> To my understanding this is simply not possible using XSD but can only be achieved using schematron. At least a reason but I a agree not necessary a good one...

I am using both, the schema and the schematron validation. It seems that the schematron does not implement any rules against the use of other/no prefixes either

**Reply by AndreasPvd** [reply]
Do you mean Schematron from CEFACT or from CEN?

**Reply by AndreasPvd** [reply]
@MartinForsberg-Ecru After re-reading the exact wording I agree to your and @phax Phillips point. Indeed, it looks like I was wrong in applying the rules in chapter 7 on the rules for instances in chapter 8. 

Many thanks for raising this question @MatthiasHainz-ceg !

**Reply by phax** [reply]
> > About validation -> To my understanding this is simply not possible using XSD but can only be achieved using schematron. At least a reason but I a agree not necessary a good one...
> 
> I am using both, the schema and the schematron validation. It seems that the schematron does not implement any rules against the use of other/no prefixes either

@MatthiasHainz-ceg Schematrons usually contain their own prefix-namespace mapping in it (element `namespace` or so). But the prefixes in the Schematrons don't need to match the Prefixes in the XML instances - they may as well differ. The namespace prefixes in Schematron only apply to the prefixes uses in the Schematron contexts and assertions

**Reply by AndreasPvd** [reply]
@phax  - Is it the same for UBL? Could I use other prefixes for cac: or the long representation as well - or is UBL here more restrictive?

**Reply by phax** [reply]
> @phax - Is it the same for UBL? Could I use other prefixes for cac: or the long representation as well - or is UBL here more restrictive?

@AndreasPvd  to the best of my knowledge you are free to choose whatever prefix you like. Best to make this your default assumption for all XML documents.

**Reply by MatthiasHainz-ceg** [reply]
> 

>   About validation -> To my understanding this is simply not possible using XSD but can only be achieved using schematron. At least a reason but I a agree not necessary a good one...

> I am using both, the schema and the schematron validation. It seems that the schematron does not implement any rules against the use of other/no prefixes either

@MatthiasHainz-ceg Schematrons usually contain their own prefix-namespace mapping in it (element namespace or so). But the prefixes in the Schematrons don't need to match the Prefixes in the XML instances - they may as well differ. The namespace prefixes in Schematron only apply to the prefixes uses in the Schematron contexts and assertions

With my very basic understanding of schematron validation, i am able to identify the namespace mapping and find rules for specific fields. I have no idea though, if it is possible to make the use of specific prefixes possible. Therefor I also could not find anything regarding this topic in the schematron itself. The [ZUGFeRD-Community](https://www.zugferd-community.net/de/infothek/technical_xml_regarding_fixed_namespace_prefixes_rsm_ram_) suggests that the prefixes are technically mandatory via the schematrons, what i could not verify myself. 

> Do you mean Schematron from CEFACT or from CEN?

I mean the EN16931-CII-validation schematron provided in this repository.


Many thanks for all the answers and the vivid discussion.
